spec.emu

```diff
@@ -139,19 +140,44 @@ markEffects: true
         1. If _lines_ is empty, then
           1. Append *undefined* to _dedented_.
         1. Else,
-          1. Let _stringElements_ be a new empty List.
+          1. Let _out_ be *""*.
           1. Let _index_ be 0.
           1. For each Record { [[String]], [[Newline]], [[LineEndsWithSubstitution]] } _line_ of _lines_, do
             1. NOTE: The first line of each block does not need to be trimmed. It's either the opening line, or it's the continuation of a line after a substitution.
             1. If _index_ = 0, let _c_ be 0; else let _c_ be _count_.
             1. Let _strLen_ be the length of _line_.[[String]].
             1. Let _trimmed_ be the substring of _line_.[[String]] from _c_ to _strLen_.
-            1. Append _trimmed_ to _stringElements_.
-            1. Append _line_.[[Newline]] to _stringElements_.
+            1. Set _out_ to the string-concatenation of _out_, _trimmed_, and _line_.[[Newline]].
             1. Set _index_ to _index_ + 1.
-          1. Let _str_ be the String value whose code units are the elements in the List _stringElements_.
-          1. Append _str_ to _dedented_.
-      1. Return CreateArrayFromList(_dedented_).
+          1. Append _out_ to _dedented_.
+      1. Return _dedented_.
+    </emu-alg>
+  </emu-clause>
+
+  <emu-clause id="sec-cook-strings-array" type="abstract operation">
+    <h1>
+      CookStringsArray (
+        _raw_: a List of Strings,
+      ): a List of either String or Undefined
+    </h1>
+    <dl class="header">
+    </dl>
+    <emu-alg>
+      1. Let _cooked_ be a new empty List.
+      1. For each element _str_ of _raw_, do
+        1. If Type(_str_) is String, then
+          1. Let _template_ be the string-concatenation of *"`"*, _str_, and *"`"*.
```

Comment:
There's no reason to do this concatenation; you can just parse the string as a |TemplateCharacters| instead of as a |NoSubstitutionTemplate|.

I don't love this implementation, though. It has weird implications for the normative behavior of `String.dedent` - for example, it rejects `{ raw: ['${'] }`. While it's true that this isn't something which can appear in a template, there's no particular reason that `String.dedent` ought to reject it. It would be better to define a more permissive version of |TemplateCharacters| (sharing most of the productions, just getting rid of ` $ [lookahead ≠ {]`  and changing the last one to `SourceCharacter but not \  or LineTerminator`), and define `TV` over the new thing. I can send a PR for that if you'd like.

(Until we get a better method for it, I'm going to use `String.dedent` on non-template strings with something like `String.dedent({ raw: ['\n' + input.replace(/\\/g, '\\\\' + '\n'] })`, and I would expect that to work even if `x` happens to contain `${` or a literal backtick.)

Reply:
> There's no reason to do this concatenation; you can just parse the string as a |TemplateCharacters| instead of as a |NoSubstitutionTemplate|.

Done.

> It would be better to define a more permissive version of |TemplateCharacters| (sharing most of the productions, just getting rid of ` $ [lookahead ≠ {]`  and changing the last one to `SourceCharacter but not \  or LineTerminator`), and define `TV` over the new thing. I can send a PR for that if you'd like.

My first attempt was to iterate all chars and look for `\` and then tried to parse the next few chars, but it quickly became unwieldy because of how many branches are needed to cover TemplateEscapeSequence and NotEscapeSequence. If you can make it work better, I'm happy to accept the PR.